pepper-apis/src/main/java/org/broadinstitute/ddp/elastic/participantslookup/ESParticipantsLookupService.java

```diff
@@ -0,0 +1,148 @@
+ *     proxy in Map(Guid, ESUsersIndexResultRow) and set it to 'proxy' element of ParticipantsLookupResultRow.</li>
+ * </ul>
+ */
+public class ESParticipantsLookupService extends ParticipantsLookupService {
+
+    private final RestHighLevelClient esClient;
+
+    public ESParticipantsLookupService(RestHighLevelClient esClient) {
+        this.esClient = esClient;
+    }
+
+    @Override
+    protected void doLookupParticipants(
+            String studyGuid,
+            String query,
+            int resultsMaxCount,
+            ParticipantsLookupResult participantsLookupResult) throws Exception {
+
+        Map<String, String> governedUserToProxy = new HashMap<>();
+        Map<String, String> governedUserToProxyExtraSearch = new HashMap<>();
+
+        var usersEsIndex = ElasticsearchServiceUtil.detectEsIndex(studyGuid, ElasticSearchIndexType.USERS);
+        var participantsEsIndex = ElasticsearchServiceUtil.detectEsIndex(studyGuid, ElasticSearchIndexType.PARTICIPANTS_STRUCTURED);
+
+        var proxiesResults = new ESUsersProxiesSearch(esClient, governedUserToProxy)
+                .setResultMaxCount(resultsMaxCount)
+                .setEsIndex(usersEsIndex)
+                .setFetchSource(USERS__INDEX__SOURCE)
+                .setQuery(query)
+                .search();
+
+        var participantsStructuredResults = new ESParticipantsSearch(esClient, governedUserToProxy, participantsLookupResult)
+                .setResultMaxCount(resultsMaxCount)
+                .setEsIndex(participantsEsIndex)
+                .setFetchSource(PARTICIPANTS_STRUCTURED__INDEX__SOURCE)
+                .setQuery(query)
+                .search();
+
+        var esResultRows = mergeResults(
+                proxiesResults, participantsStructuredResults, governedUserToProxy, governedUserToProxyExtraSearch);
+
+        // if any extra-proxies found then fetch their data from index "users", add to proxiesResults, and merge results again
+        if (governedUserToProxyExtraSearch.size() > 0) {
+            esResultRows = reMergeResultsWithExtraFoundProxies(
+                    query, resultsMaxCount, governedUserToProxy, governedUserToProxyExtraSearch,
+                    usersEsIndex, proxiesResults, participantsStructuredResults);
+
+        }
+
+        participantsLookupResult.setResultRows(esResultRows);
+    }
+
+    /**
+     * Adds to results of "participants_structured" index the 'proxies' data results (taken from
+     * "users" index results).
+     */
+    private List<ParticipantsLookupResultRow> mergeResults(
+            Map<String, ESUsersIndexResultRow> usersResults,
+            Map<String, ESParticipantsStructuredIndexResultRow> participantsStructuredResults,
+            Map<String, String> governedUserToProxy,
+            Map<String, String> governedUserToProxyExtraSearch) {
+        List<ParticipantsLookupResultRow> resultList = new ArrayList<>();
+        for (var participant : participantsStructuredResults.values()) {
```

Comment:
I don't see anywhere that addresses this scenario: what if there are matches in the first search for proxies, but then the second search in the participants index does not return the governed user from the first search? It might be possible that the `query` matched a bunch of non-governed users, and because of `resultsMaxCount` some of the governed users from the first search is not returned.

But I think the approach you have here is slightly different and I think it'll be fine. If I'm understanding, the main result set of final response will come from this second search on the participants index. If the search returned a governed user, then great it'll be included in final response along with the proxy that we found in the first search. If not, then the proxies we found in the first search will simply be left out. I think this is fine.

Reply:
> and because of resultsMaxCount some of the governed users from the first search is not returned
Yes, it's possible. I can split search in "participants" into 2 runs:
1) search for governedUsers (for proxies found in "users").
1) search for participants by "query" (coming from frontend), limited by `resultsMaxCount`.
This search result consider as higher priority (than proxies found in "users"). Therefore take the results of this search (+ proxy users of the governedUsers existing in this list). If this result >= `resultsMaxCount` then take only this list as a final result. 
If it < `resultsMaxCount` then find in proxies list (found in `users`) the proxies which not exist in the 1st result, and find governed users for them (also checking that the extra amount together with 1st list will be <= `resultsMaxCount`).

Reply:
So, that priority of finding among participants data (first,last name, email, guid, hruid) will be higher than among proxies (first, last name, email, guid, hruid).

Reply:
I don't see how the 2 runs you described is any better than what you have now. For case 2 of `< resultsMaxCount`, if this is the case, then that means we have room to search for the governed users of proxies. But isn't that exactly what you get back now in your current approach? If the `query` doesn't find enough matches to fill `resultsMaxCount`, then the rest of the matches you get back should be the governed users that we're looking for?

Reply:
I think the way you have it now is fine.

Reply:
Let me double-check..
Currently it works like this:
1) Search in "users" (in proxy users only) by specified "query" (in name/email/guid/hruid/legacyId...).
2) Search in "participants" by compound search: governed users (by guids detected from 1st seaech in "users") + by "query" (in name/email/guid/hruid/invitations/legacyId...). The total result contains governed users detected from proxies (found in 1st search) + governed users AND normal users found by "query". 
And here can happen that the total list exceeds `resultsMaxCount` and some of governed users/normal users (found by "query") excluded from result.. 
But ok, I think it is not so bug issue probably..